Use cdk-virtual-scroll for long cipher lists

eliykat opened the pull request:
## Objective

Boost performance on large vaults (see e.g. #1597) by reducing the number of elements rendered at once. (This replaces a previous PR which used a different library which wasn't working out for me.)

Currently we have paging logic, which loads the first 100 ciphers, then the next 100, etc, as you scroll down. However:
* this prevents the user from accessing all their ciphers at once
* performance will degrade as more items load in
* this is not implemented in all areas of the app (in particular, search results)
* the `ngx-infinite-scroll` library used for this purpose does not work with Ivy, which has prevented us switching over to it (see #1412).

Here's an example of how you can break it currently (I'm doing this very intentionally, but it can happen). Note that the extension fails to re-open afterwards, I had to restart Chrome. [Video](https://user-images.githubusercontent.com/31796059/119082776-181b9200-ba42-11eb-854f-d41dce2198c6.mov)

Instead we can use [cdk-virtual-scroll](https://material.angular.io/cdk/scrolling/overview) from the Angular CDK.

> The <cdk-virtual-scroll-viewport> displays large lists of elements performantly by only rendering the items that fit on-screen. Loading hundreds of elements can be slow in any browser; virtual scrolling enables a performant way to simulate all items being rendered by making the height of the container element the same as the height of total number of elements to be rendered, and then only rendering the items in view. Virtual scrolling is different from strategies like infinite scroll where it renders a set amount of elements and then when you hit the end renders the rest.

Here's how it performs after these changes: [Video](https://user-images.githubusercontent.com/31796059/119082780-1b168280-ba42-11eb-8933-c37c07e3196e.mov)

✅  better performance
✅  easier to implement across different parts of the app
✅  no external dependencies
✅  compatible with Angular Ivy (I tested `dist` builds in Chrome, FF and Safari to make sure - but I would also expect this given that it's part of the Angular CDK)

# Code changes

* `ciphers-list.component` has been refactored into `cipher-row.component`. 
  * `ciphers-list` used `*ngFor` to render a list of ciphers, so the looping logic was inside the component
  * `ciphers-row` removes the looping logic from this component and just renders a single cipher. This lets us use a mix of "real" and virtual scrolling in different parts of the app, because the looping logic is lifted up to the parent component
* `base.scss` - where virtual scroll is used, the scrollbar is on the `cdk-virtual-scroll-viewport` element, not the `content` element. Apply the scrollbar styles and some other miscellaneous styling to that element.
* `popup-utils.service` - this has some helpers used to save and restore the scrolling position. Tweaked so that we can specify a different scrolling container other than `content`.
* various component templates: use virtual scroll in `groupings.component` for the favorites list and search results, and in `ciphers.component` for the ciphers list. Use regular `*ngFor` in all other locations that used `ciphers-list.component`.
* `tsconfig.json` - enable Ivy.
* remove `ngx-infinite-scroll` dependency

## Other clients
* desktop - should be trivial to implement virtual scroll here as well - but I'll wait for confirmation before doing so
* web - a major limitation of virtual scroll is that it can't refer to a parent element's scrollbar. This means that the list of ciphers in the web vault would have its own scrollbar, as well as the scrollbar on the page, which wouldn't be workable. So I propose to leave web alone - although for what it's worth, it appears to have paging logic implemented throughout more consistently. I still don't love it but I think it's less of a problem compared to the browser client.

madhanhere commented on src/popup/app.module.ts:
```diff
@@ -1,6 +1,6 @@
 import { DragDropModule } from '@angular/cdk/drag-drop';
+import { ScrollingModule } from '@angular/cdk/scrolling';
```
align the imports

madhanhere commented on src/popup/vault/current-tab.component.html:
```diff
@@ -27,9 +27,9 @@
-                <app-ciphers-list [ciphers]="loginCiphers" title="{{'autoFill' | i18n}}" [showView]="true"
-                    (onSelected)="fillCipher($event)" (onView)="viewCipher($event)" *ngIf="loginCiphers.length">
-                </app-ciphers-list>
+                <app-cipher-row *ngFor="let c of loginCiphers" [cipher]="c" title="{{'autoFill' | i18n}}" [showView]="true"
```
Just a suggestion instead using the variable name `c` we can use `loginChiper` and I can see few other places below variable name `c` is used there we can use appropriate name.

cscharf replied:
What he said.

madhanhere commented on src/popup/vault/groupings.component.html:
```diff
@@ -23,131 +23,135 @@
-            <div class="box-header">
-                {{'noneFolder' | i18n}}
-                <div class="flex-right">{{noFolderCiphers.length}}</div>
+            <div class="box list" *ngIf="nestedFolders && nestedFolders.length">
```
we can use the safe navigation operator (?) to make it simple `*ngIf="nestedFolders?.length"`

cscharf replied:
Actually, this should not be using falsey logic but explicit logic.

```suggestion
            <div class="box list" *ngIf="nestedFolders != null && nestedFolders.length > 0">
```

eliykat replied:
Good point

eliykat commented:
I've fixed the few style issues that were pointed out. I've also removed the use of `virtualScroll` for the favorites list, as this should not reasonably have enough items to cause any slowdown, so `virtualScroll` would probably just add overhead there.

cscharf reviewed (commented):
I think this looks good Thomas, however I'm adding Oscar on here as well for a quick review/sanity check if there's anything I missed.

Hinton reviewed (approved):
Looks good to me, but I'm not quite familiar with the previous unlimited scroll.
